### ISSUE TITLE
Canu sensitivity to genomeSize parameter

### ISSUE THREAD

**Issue by jpummil** [marbl/canu#260]
Greetings!

I am currently working with some Bio faculty on a de novo assembly of Centruroides vittatus (Striped Bark Scorpion).  One difficulty I seem to be having involves the command line parameter "genomeSize=" in Canu. Of course, one of the real difficulties with de novo is you likely don't know the size of the genome. You may not even have an estimate from a close relative species.

I currently have two PacBio runs (different specimens in each) labeled Q1133 and Q1171. It seems that by varying the "genomeSize=" parameter, the output of the assembly can be significantly different! I am attaching a Quast output for some recent runs...two for each data set, with only the genomeSize parameter modified (genomeSize=170 and genomeSize=350). 

It seems that not many people have produced tools to estimate genome size in new projects that involve PacBio or ONT data...I did have some old Illumina data that was kind of iffy, but I tried the new GenomeScope tool from Schatz, et al and it comes up pretty consistently around 180Mb for an expected genome size. It was suggested to try the tool using the corrected reads out of Canu, but the tool doesn't seem to be able to evaluate that sort of input at this time.

Any suggestions on how to best approach the problem (if it IS a problem and not simply an error on my part)? I am using essentially default values in Canu with the exception of the "ovsMemory=" parameter as I was hitting the error seen by others.

[report.pdf](https://github.com/marbl/canu/files/493478/report.pdf)

**Comment by skoren** [comment]
The only things genomeSize will affect is how many reads are selected for consensus and assembly (40X * genome size) and the NG statistics canu reports in its logs. So it is possible that by increasing the parameter you get slightly more corrected reads and a larger assembly size, especially if you had low coverage from the correction in the 170M case. You can test this by keeping the same genome size and increasing corOutCoverage=80 (from the default of 40) which will generate more corrected reads.

The 170 and 350 assemblies do look quite similar for Q1133, I would check how much coverage  in corrected reads you had for both Q1133 runs (it should be output in the canu-scripts/canu.\*.out logs). As for Q1171 the difference is very large but the small genome size for the 350M size makes me think that assembly failed somewhere or didn't run correctly. There is a bug in Canu 1.3 in one step where if an assembly stops and you resume it manually, it could continue on without realizing it hadn't completed. If you have the canu-scripts folder from that run, we can check if there are any errors in the output.

**Comment by jpummil** [comment]
Greetings Sergey! Thank you for the information, I believe it is most helpful. 

I am using a recent git pull, so software should be current. I did still have the directory structure intact for the runs, but looking in the canu-scripts directory in each run shows nothing. Perhaps due to the output being routed to an output file by the scheduler? I do have all four that are attached to this post.

Happy to try the corOutCoverage=80 suggestion as well. May take a day or so to get a node on the cluster again as it's currently pretty busy.



[Output.zip](https://github.com/marbl/canu/files/498998/Output.zip)




 of those and they seem to have lots of metric data in them (attached)....

**Comment by SchwarzEM** [comment]
skoren wrote: "The only things genomeSize will affect is how many reads are selected for consensus and assembly (40X * genome size)".

Does this mean that Canu uses *all* of the error-corrected reads for genome assembly?   In other words, if one changes corOutCoverage from its default of 40 to, say, corOutCoverage=80, does one end up assembling one's genome with all 80x of the corrected-read coverage?

My reason for asking is that PBcR-MHAP, Canu's predecessor, would error-correct 40x coverage (by default) but then would assemble only the longest 25x (by default).  In running PBcR-MHAP, if one wanted to increase the coverage of reads that one used in the *assembly* step, it was necessary to provide an altered parameter, typically in the .spec file, such as "assembleCoverage = 35".

I couldn't find any sign of a 'assembleCoverage' parameter in Canu.  Is it just that I didn't know where to look?

**Comment by SchwarzEM** [comment]
OK, answering my own question, it looks as if genomeSize *does* simultaneously specify both the coverage of reads to be corrected and the coverage of reads to be subsequently assembled.  I.e., one is identical to the other.  That's helpful to know!

http://canu.readthedocs.io/en/latest/parameter-reference.html#read-correction

**Comment by jpummil** [comment]
A bit of an update on modifying corOutCoverage from its default of 40 to to 80. There were differences (see quast file attached), but not significant. Column 1 is at the default of 40, column 2 is identical , but added corOutCoverage=80 as Sergey suggested...
[quast_report.zip](https://github.com/marbl/canu/files/519879/quast_report.zip)

**Comment by skoren** [comment]
Sorry for not replying for a bit. I looked through your outputs and realized there is one other thing that genomeSize will affect is the estimate of input coverage which automatically adjusts some correction parameters. In your case, since you're doubling the genome size, the parameters for the 350MB runs are sensitive mode while the 170 are not. 

I think you should generally be OK with the 170 genome size assemblies. In most cases, your output seems to have sufficient corrected data for assembly. You could try genomeSize=170 with sensitive parameters (```corMhapSensitivity=high corOutCoverage=0 errorRate=0.035```) for comparison. Canu will also try to split haplotypes in your assembly so if the sample is very heterozygous you can likely end up with double the genome size assembled (or more if your DNA came from a population and not a single individual).

**Comment by jpummil** [comment]
Thanks for the additional tips, Sergey! I just submitted a run to see what the additional sensitivity parameters will yield. 

Interesting to note that when I tried to combine data during a Canu assembly from two different specimens, I did in fact see the doubling of size effect that you mentioned! I hope to have a better idea of where I stand with these assemblies soon as I'd like to try and apply Quiver/Arrow to the preferred one and see how much, if any, improvements can be had there as well.

Cheers,
